File: sections/65_polynomial_algorithms_recurrences/problems/a_cyclic_agreement/solution.cpp
#include <bits/stdc++.h>
#include "../../ntt.hpp"
using namespace std;
using course_ntt::MOD;
using course_ntt::convolution;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    string s, t;
    if (!(cin >> n >> s >> t)) return 0;

    // Equality of two two-bit symbols is
    // (1+xX)(1+yY)/4 = (1+xX+yY+xyXY)/4 for signs +/-1.
    const int inverse_four = course_ntt::power(4, MOD - 2);
    const string alphabet = "ACGT";
    vector<int> score(n, n);
    for (int feature = 0; feature < 3; ++feature) {
        vector<int> left(n), doubled_right(2 * n);
        for (int i = 0; i < n; ++i) {
            int a = alphabet.find(s[i]);
            int b = alphabet.find(t[i]);
            int sign_a = feature == 0 ? (a & 1) : feature == 1 ? (a >> 1) : ((a & 1) ^ (a >> 1));
            int sign_b = feature == 0 ? (b & 1) : feature == 1 ? (b >> 1) : ((b & 1) ^ (b >> 1));
            left[n - 1 - i] = sign_a ? MOD - 1 : 1;
            int value = sign_b ? MOD - 1 : 1;
            doubled_right[i] = doubled_right[i + n] = value;
        }
        vector<int> product = convolution(left, doubled_right);
        for (int shift = 0; shift < n; ++shift) {
            score[shift] += product[n - 1 + shift];
            if (score[shift] >= MOD) score[shift] -= MOD;
        }
    }
    int best = -1, best_shift = 0;
    for (int shift = 0; shift < n; ++shift) {
        int matches = static_cast<int>(1LL * score[shift] * inverse_four % MOD);
        if (matches > best) best = matches, best_shift = shift;
    }
    cout << best << ' ' << best_shift << '\n';
}

File: sections/65_polynomial_algorithms_recurrences/problems/a_cyclic_agreement/solution.py
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
from ntt import MOD, convolution


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    s, t = data[1].decode(), data[2].decode()
    code = {char: index for index, char in enumerate("ACGT")}
    scores = [n] * n
    for feature in range(3):
        left = [0] * n
        right = [0] * (2 * n)
        for i, (x, y) in enumerate(zip(s, t)):
            a, b = code[x], code[y]
            bit_a = (a & 1) if feature == 0 else (a >> 1) if feature == 1 else ((a & 1) ^ (a >> 1))
            bit_b = (b & 1) if feature == 0 else (b >> 1) if feature == 1 else ((b & 1) ^ (b >> 1))
            left[n - 1 - i] = MOD - 1 if bit_a else 1
            value = MOD - 1 if bit_b else 1
            right[i] = right[i + n] = value
        product = convolution(left, right)
        for shift in range(n):
            scores[shift] = (scores[shift] + product[n - 1 + shift]) % MOD
    inverse_four = pow(4, MOD - 2, MOD)
    matches = [value * inverse_four % MOD for value in scores]
    best = max(matches)
    print(best, matches.index(best))


if __name__ == "__main__":
    main()
